src/Obelisk/Frontend/GoogleAnalytics.hs

```diff
@@ -1,54 +1,250 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wall #-} -- -Werror #-}
 
 module Obelisk.Frontend.GoogleAnalytics
   ( googleAnalyticsFromConfig
   , googleAnalytics
   , trackingIdPath
   , getTrackingId
+  , Analytics(..)
+  , GoogleAnalyticsT(..)
+  , GtagJSCall(..)
+  , gaClickEvent
+  , gaOutboundClickEvent
+  , runGoogleAnalyticsT
   ) where
 
-import Control.Monad (void)
+import Control.Monad
+import Control.Monad.Exception
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Primitive
+import Control.Monad.Ref
+import Control.Monad.State.Strict
 import Data.Text (Text)
-import GHCJS.DOM.Types (liftJSM)
 import Language.Javascript.JSaddle.Evaluate (eval)
 import Obelisk.Configs
 import Obelisk.Route.Frontend
 import Reflex.Dom.Core
 import qualified Data.Text as T
+import qualified Data.Aeson as Aeson
+import Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+import Language.Javascript.JSaddle
+import Reflex.Host.Class
+
+import qualified Data.HashMap.Strict as HMap
+
+class Monad m => Analytics t w m | m -> w where
+  tellAnalytics :: Event t w -> m ()
+  -- ^ Tell analytics something, with asynchronous delivery.
+
+  -- tellAnalyticsSync :: Event t w -> m ()
+  -- ^ Tell analytics something, with synchronous delivery.
+  --   This waits until the message is delivered before returning.
+
+  -- tellAnalyticsWithSync :: Event t w -> m (STM Bool)    <- this type is conceptually flawed
+  -- ^ Tell analytics something, with asynchronous delivery.
+  --   Returns a 'STM' action that becomes 'True' after delivery of this particular message.
+  --   This is the preferred way to implement synchronous delivery with timeout.
+
+  -- syncAnalytics :: m ()
+  -- ^ Synchronize on the delivery of every message outstanding at the time of the call.
+  --   This returns after all such messages have been delivered.
+
+  -- asyncAnalytics :: m (STM Bool)
+  -- ^ Returns an 'STM' action that becomes 'True' after all currently outstanding messages have been delivered.
+  --   This is an asynchronous variant of 'syncAnalytics', and the preferred way of doing a "global" sync with timeout.
+
+data GtagJSCall = GtagJSCall
+  { _gtagJSCall_method :: !Text
+  , _gtagJSCall_action :: !Text
+  , _gtagJSCall_params :: !Aeson.Value
+  }
```

Comment:
Let's derive at least a few things here `(Generic, Eq, Ord, Show)` seem like ok default set

Reply:
That's plenty ok by me; I was meaning to do it, but got overlooked turned out not to be immediately neccessary.